default/python/AI/freeorion_tools/_freeorion_tools.py

```diff
@@ -218,3 +219,54 @@ def get_partial_visibility_turn(obj_id):
     """
     visibility_turns_map = fo.getUniverse().getVisibilityTurnsMap(obj_id, fo.empireID())
     return visibility_turns_map.get(fo.visibility.partial, -9999)
+
+
+class ReadOnlyDict(Mapping):
```

Comment:
Looks like you don't use Mapping and it can be replaced to object.

Reply:
Mapping is used to provide dict-like utilities such as ```__contains__```, ```keys```, ```values```, ```items```, ```get```. It also allows copying like ```dict(ReadOnlyDict)```.

Reply:
> Mapping is used to provide dict-like utilities such as __contains__

You have own `__contains__`. 

> It also allows copying like dict(ReadOnlyDict).

No, you don't use `Mapping.__iter__`


Mapping docs say:
```
"""A Mapping is a generic container for associating key/value
    pairs.

    This class provides concrete generic implementations of all
    methods except for __getitem__, __iter__, and __len__.
"""
```
And I see much more methods redefined.

Reply:
> You have own __contains__.

Morlic may not have chosen the best examples to mention, but Mapping does still provide other methods that seem worthwhile.  

> And I see much more methods redefined.

__contains__ is the only one I noticed.   @Morlic-fo  I'll admit I'm left wondering why it is redefined-- is there a good reason, to put into a comment, or did you just get a bit carried away?

Reply:
You are right that ```__contains__``` is already defined in mapping, no reason to redefine. ```__nonzero__``` was redundant as well as python standards dictate that, if no ```__nonzero__``` function is defined, ```__len__``` is called instead.
I removed those function definitions.